src/thunderbots/software/util/parameter/parameter.h

```diff
@@ -13,121 +25,199 @@
+
+            Parameter<T>::registerParameter(std::make_shared<Parameter<T>>(*this));
+        }
+
+        /**
+         * Returns the global path in the ROS parameter server where this parameter is stored
+         *
+         * @return the global path in the ROS parameter sever where this parameter is stored
+         */
+        const std::string getROSParameterPath() const
+        {
+            return NamespaceForParameters + "/" + name();
+        }
+
+        /**
+         * Returns the value of this parameter
+         *
+         * @return the value of this parameter
+         */
+        const T value() const
+        {
+            // get the value from the parameter in the registry
+            if (Parameter<T>::getMutableRegistry().count(this->name_)){
+                auto& param_in_registry = Parameter<T>::getMutableRegistry().at(this->name_);
+                return param_in_registry->value_;
+            }
+            // if the parameter hasn't been registered yet, return default value
+            return this->value_;
+        }
+        /**
+         * Returns the name of this parameter
+         *
+         * @return the name of this parameter
+         */
+        const std::string name() const
+        {
+            return name_;
+        }
+
+        /**
+         * Updates the value of this Parameter with the value from the ROS
+         * Parameter Server
+         */
+        void updateValueFromROSParameterServer()
+        {
+            ros::param::get(getROSParameterPath(), this->value_);
+        }
+
+        /**
+         * Returns a reference to the Parameter registry. The registry is a list of
+         * pointers to all the existing Parameters.
+         *
+         * @return An immutable reference to the Parameter registry
+         */
+        static const std::map<std::string, std::shared_ptr<Parameter<T>>>& getRegistry()
+        {
+            return Parameter<T>::getMutableRegistry();
+        }
+
+        /**
+         * Returns a reference to the config struct. The config struct contains
+         * all the current configurations
+         *
+         * @return An immutable reference to the Config struct
+         */
+        static const dynamic_reconfigure::Config& getConfigStruct()
+        {
+            return Parameter<T>::getMutableConfigStruct();
+        }
+
+        /**
+         * Registers (adds) a Parameter to the registry. Since the shared pointer is moved
+         * into the registry, the pointer may not be accessed by the caller after this
+         * function has been called.
+         *
+         * Also registers params to the static configuration struct used to
+         * set parameters
+         *
+         * @param parameter A shared pointer to the Parameter to add. This pointer may not
+         * be accessed by the caller after this function has been called.
+         */
+        static void registerParameter(std::shared_ptr<Parameter<T>> parameter)
```

Comment:
Question: Since the shared_ptr can't be used by the caller after calling this function, could we make this accept a unique_ptr instead? This would likely require changing how the caller creates/uses the pointer too, so might not be possible if we really rely on the shared_ptr before calling this function.

The benefit would be that the unique_ptr would have to be passed using `std::move`, and makes the code more explicit that the pointer is being moved and can no longer be accessed after using it to call this function. This is mostly a nitpick on making the code very explicit about the purpose and use of the pointer.

Reply:
Yeah thats true, changed it to `unique_ptr` and used `std::move` again. It definitely makes it more explicit